src/App.tsx

```diff
@@ -1,61 +1,128 @@
+import { ChangeEvent, useState } from 'react';
+import { TodoList } from './components/TodoList';
+import { Todo } from './types/Todo';
 import './App.scss';
 
-// import usersFromServer from './api/users';
-// import todosFromServer from './api/todos';
+import usersFromServer from './api/users';
+import todosFromServer from './api/todos';
+import { getUserById, getNextId } from './helpers';
+
+const refactoredTodos: Todo[] = todosFromServer.map((todo) => ({
+  ...todo,
+  user: getUserById(todo.userId),
+}));
 
 export const App = () => {
+  const [todos, setTodos] = useState(refactoredTodos);
+  const [title, setTitle] = useState('');
+  const [selectedUser, setSelectedUser] = useState(0);
+  const [titleValid, setTitleValid] = useState(true);
+  const [userValid, setUserValid] = useState(true);
+
+  const handleTitleChange = (event: ChangeEvent<HTMLInputElement>) => {
+    setTitle(event.target.value);
+    setTitleValid(Boolean(event.target.value));
+  };
+
+  const handleUserChange = (event: ChangeEvent<HTMLSelectElement>) => {
+    setSelectedUser(+event.target.value);
+    setUserValid(Boolean(event.target.value));
+  };
+
+  const handleFormSubmit = (event: React.FormEvent<HTMLFormElement>) => {
+    event.preventDefault();
+
+    setTitleValid(Boolean(title.trim()));
+    setUserValid(Boolean(selectedUser));
+
+    if (!selectedUser || !title.trim()) {
+      return;
+    }
+
+    setTodos((currentTodos) => {
+      const nextId = getNextId(currentTodos);
+      const newUser = getUserById(selectedUser);
+
+      return [
+        ...currentTodos,
+        {
+          id: nextId,
+          title,
+          completed: false,
+          userId: newUser ? newUser.id : null,
+          user: newUser,
+        },
+      ];
```

Comment:
Let's create new todo as new variable and just add it to array in setter to improve readability

```suggestion
   const newTodo = {
          id: getNextId(currentTodos),
          title,
          completed: false,
          userId: selectedUser? selectedUser : null,
          user: getUserById(selectedUser),
    };
 
    setTodos((currentTodos) => [...currentTodos, newTodo]);
```